lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -953,6 +980,27 @@ def _upgrade_opensearch(self, event: _UpgradeOpenSearch) -> None:
                 event.defer()
                 return
 
+        # TODO upgrade: catch http errors
+
+        # TODO upgrade: does this affect user set values?
+        # TODO look at `preserve_existing` https://www.elastic.co/guide/en/elasticsearch/reference/8.13/indices-update-settings.html
+        # Increase timeout before shard allocation replicates shards on offline (e.g. upgrading)
+        # node to other nodes
+        # Used instead of disabling shard allocation entirely
+        # (Less dangerous if this unit fails to upgrade & reset the timeout—it's safer to have
+        # increased timeout than shard allocation disabled entirely, especially for long-running
+        # upgrades)
+        # Replaces "Disable shard allocation" step from
+        # https://www.elastic.co/guide/en/elastic-stack/8.13/upgrading-elasticsearch.html#rolling-upgrades
+        self.opensearch.request(
+            "PUT",
+            "/_all/_settings",
+            payload={"settings": {"index.unassigned.node_left.delayed_timeout": "5m"}},
+        )
```

Comment:
You're right! What we can do here is: 
1. fetch the current setting value 
2. override it if it's < 5min, or keep as is otherwise
3. set it back to the original value if changed

Reply:
The idea I had was more with regards to the relation data. Not the `StoredState`.

---

Overriding it - without setting it back - is certainly inconvenient for the user who may have to reconfigure their indices settings, but definitely safer than disabling replica shards allocation in the case of a **long running** upgrade - which the Elastic docs doesn't really take into consideration.

I believe that was the conclusion we reached when we discussed the pros and cons of each. The major con of disabling replica shards allocation (as advised by the docs) is in the case of long upgrades - where no new replica shards **(of new indices)** will be created. (existing replicas will still be replicated)

---

I just made a research now, and I believe we must differentiate between a **cluster wide** `index.unassigned.node_left.delayed_timeout` and an index-level `delayed_timeout`.
The former only overrides the ones in the indices where no `delayed_timeout` has been set, **NOT** where it has been set.

---

Meaning that if we wanna go down this route, we should:
1. fetch the shards located in the node about to upgrade
2. retrieve the index names they belong to 
3. retrieve the delayed timeout for each - store it in the peer data - override it with `5m` (if 5m greater than whatever is set in there)
4. after upgrade reset it based on the relation data

---

Makes sense?

Reply:
sounds good

Reply:
it is the default behavior of the call you made above - it only affects the indices not explicitely configured by the user

Reply:
I gave it some more thoughts, we should be selective, please refer to [here](https://github.com/canonical/opensearch-operator/pull/261#discussion_r1583015342)